hazelcast/src/main/java/com/hazelcast/map/impl/BasicMapContextQuerySupport.java

```diff
@@ -142,9 +142,10 @@ public Set queryLocalMember(String mapName, Predicate predicate, IterationType i
             if (partitionIds.isEmpty()) {
                 return result;
             }
-        } catch (QueryResultSizeExceededException e) {
-            throw ExceptionUtil.rethrow(e);
         } catch (Throwable t) {
+            if (t.getCause() instanceof QueryResultSizeExceededException) {
```

Comment:
It's not the problem that I'm not encapsulating my custom exception. It's the other way around. My custom exception got wrapped with a `ExecutionException`, so I didn't catch it properly.

Originally there was no `ExceptionUtil.rethrow(t);` at all. It was part of the fallback mechanism of the query logic to ignore the first exceptions at all. This is how the code looked in the (mostly duplicated) query logic through the system:
```java
try {
  queryMembers();
} catch (Throwable t) {
  // ignore or at most log
}
try {
  queryPartitions()
} catch (Throwable t) {
  ExceptionUtil.rethrow(t);
}
```
The problem with my `QueryResultSizeExceededException` is that we have to catch it after the `queryMember()` part. Otherwise we will do the whole query work again for each partition (and with the same exception being thrown again).

So what is actually implemented is a catch of the `QueryResultSizeExceededException` with an immediate rethrow:
```java
try {
  queryMembers();
} catch (QueryResultSizeExceededException e) {
  ExceptionUtil.rethrow(e);
} catch (Throwable t) {
  // ignore or at most log
}
try {
  queryPartitions()
} catch (Throwable t) {
  ExceptionUtil.rethrow(t);
}
```
The problem is, that this didn't work, because the `QueryResultSizeExceededException` was always encapsulated in an `ExecutionException`. Because of this I had to move the rethrow in the `catch (Throwable)` and look for the cause.